### PR TITLE
Fix comparisons for abstract specs

### LINKED ISSUE #20310: Bug: comparisons between abstract specs can raise -> spack spec callpath fails
I installed Spack v0.16, and I tried to install callpath. I get the error message below.

I discussed the issue with @becker33, and he says it looks like there is a bug in the dyninst package and the elf providers, so it fails if there is a dependance on dyninst and elf. It looks like the error is releated to `elfutils`.

The error does not show up in Spack v0.15.

### Steps to reproduce the issue

Install Spack, and checkout v0.16

```console
$ spack install callpath
```

### Error Message

```console
Error: '<' not supported between instances of 'str' and 'NoneType'
```

### Information on your system

I run on CentOS 7.

### Additional information

- [ ] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by becker33**
Fixes #20310

### CODE REVIEW

File: lib/spack/spack/test/spec_syntax.py [hunk 1]
```diff
@@ -806,3 +806,26 @@ def test_kv_with_spaces(self):
     ])
     def test_target_tokenization(self, expected_tokens, spec_string):
         self.check_lex(expected_tokens, spec_string)
+
+    @pytest.mark.regression('20310')
+    def test_compare_abstract_specs(self):
```
Comment (scheibelp) [inline lib/spack/spack/test/spec_syntax.py:811]: This test appears to pass for me without applying the changes that this PR makes to `spec.py` on bb9f5d61 (a recent-ish commit on develop)

Comment (becker33) [reply lib/spack/spack/test/spec_syntax.py:811]: @scheibelp I just rebased on the most recent develop and ran the tests without the changes to `spec.py`.

```
$ spack unit-test -k compare_abstract lib/spack/spack/test/spec_syntax.py
=================================== test session starts =================================
platform linux -- Python 3.7.2, pytest-3.2.5, py-1.4.34, pluggy-0.4.0
rootdir: /g/g0/becker33/dev-spack/spack, inifile: pytest.ini
collected 47 items                                                                                                                                                  

lib/spack/spack/test/spec_syntax.py F
=========================== short test summary info ===============================
FAIL lib/spack/spack/test/spec_syntax.py::TestSpecSyntax::()::test_compare_abstract_specs

========================= slowest 20 test durations ==============================
0.00s setup    lib/spack/spack/test/spec_syntax.py::TestSpecSyntax::test_compare_abstract_specs
0.00s call     lib/spack/spack/test/spec_syntax.py::TestSpecSyntax::test_compare_abstract_specs
0.00s teardown lib/spack/spack/test/spec_syntax.py::TestSpecSyntax::test_compare_abstract_specs
======================================= FAILURES ===================================
______________________ TestSpecSyntax.test_compare_abstract_specs ___________________________

self = <spack.test.spec_syntax.TestSpecSyntax object at 0x2aaab22c0b38>

    @pytest.mark.regression('20310')
    def test_compare_abstract_specs(self):
        """Spec comparisons must be valid for abstract specs.
    
            Check that the spec cmp_key appropriately handles comparing specs for
            which some attributes are None in exactly one of two specs"""
        # Add fields in order they appear in `Spec._cmp_node`
        constraints = [
            None,
            'foo',
            'foo.foo',
            'foo.foo@foo',
            'foo.foo@foo+foo',
            'foo.foo@foo+foo arch=foo-foo-foo',
            'foo.foo@foo+foo arch=foo-foo-foo %foo',
            'foo.foo@foo+foo arch=foo-foo-foo %foo cflags=foo',
        ]
        specs = [Spec(s) for s in constraints]
    
        for a, b in itertools.product(specs, repeat=2):
            # Check that we can compare without raising an error
>           assert a <= b or b < a

lib/spack/spack/test/spec_syntax.py:831: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

s = , o = foo

>   lambda s, o: o is not None and s._cmp_key() <= o._cmp_key())
E   TypeError: '<=' not supported between instances of 'NoneType' and 'str'

lib/spack/llnl/util/lang.py:251: TypeError
=================================== 46 tests deselected ====================================
========================== 1 failed, 46 deselected in 0.22 seconds ===============================
```

Comment (becker33) [reply lib/spack/spack/test/spec_syntax.py:811]: Looks like it's a python2/3 thing, thanks for the suggestion @scheibelp 
```
$ python2 `which spack` unit-test -k compare_abstract lib/spack/spack/test/spec_syntax.py
================================= test session starts =================================
platform linux2 -- Python 2.7.16, pytest-3.2.5, py-1.4.34, pluggy-0.4.0
rootdir: /g/g0/becker33/dev-spack/spack, inifile: pytest.ini
collected 47 items                                                                     

lib/spack/spack/test/spec_syntax.py .

============================== slowest 20 test durations ==============================
0.05s setup    lib/spack/spack/test/spec_syntax.py::TestSpecSyntax::test_compare_abstract_specs
0.00s call     lib/spack/spack/test/spec_syntax.py::TestSpecSyntax::test_compare_abstract_specs
0.00s teardown lib/spack/spack/test/spec_syntax.py::TestSpecSyntax::test_compare_abstract_specs
================================= 46 tests deselected =================================
======================= 1 passed, 46 deselected in 0.09 seconds =======================

**Review by scheibelp — APPROVED** [review]

**Comment by becker33** [comment]
@scottwittenburg is this `stc` bug in the pipeline a known issue or something new with this PR?

**Comment by scottwittenburg** [comment]
It seems to have just appeared in the last 12 hours or so.  I see that failure in a develop pipeline from 4 hours ago, so I suspect it's one of those that got in because the gitlab pipeline either failed and it was assumed a false positive, or because the pipeline failure status had not yet been reported on the github PR (either because of the huge backlog, or because of the gitlab bug that prevents some pipelines from ever being marked as finished).

At any rate, I don't think it was introduced in this PR, thank you for checking though 😄

**Pull request merged**